Indexing/CodeTree.hpp

```diff
@@ -254,65 +256,70 @@ class CodeTree
     CodeOp* _alternative;
   };
 
+  /**
+   * A search structure that collects alternatives of the same
+   * kind (either function symbols or ground terms) and offers
+   * more efficient searching and insertion over them.
+   */
   struct SearchStruct
   {
-    void destroy();
-    bool isFixedSearchStruct() const { return true; }
-
+    /**
+     * Fills out pointer @b tgt where @b insertedOp should be
+     * (or is) inserted in the structure. If @b doInsert is true
+     * an entry is inserted if not found.
+     */
+    template<bool doInsert>
     bool getTargetOpPtr(const CodeOp& insertedOp, CodeOp**& tgt);
 
+    /**
+     * Returns code op in the structure matching the content
+     * of flat term entry @b ftPos.
+     */
     CodeOp* getTargetOp(const FlatTerm::Entry* ftPos);
+    inline size_t length() const { return targets.size(); }
 
     enum Kind
     {
       FN_STRUCT,
       GROUND_TERM_STRUCT
     };
 
+    /**
+     * Actual code op for this search structure. This construction
+     * implies that search structure operations cannot be stored
+     * in @b CodeBlock containers.
+     */
     CodeOp landingOp;
     Kind kind;
+    vvector<CodeOp*> targets;
+
   protected:
-    SearchStruct(Kind kind);
+    SearchStruct(Kind kind, size_t length);
   };
 
-  struct FixedSearchStruct
+  template<SearchStruct::Kind k>
+  struct SearchStructImpl
   : public SearchStruct
   {
-    FixedSearchStruct(Kind kind, size_t length);
-    ~FixedSearchStruct();
+    USE_ALLOCATOR(SearchStructImpl);
```

Comment:
Did this give a performance win? If not, no need for it any more.

Reply:
Not sure, I will remove it.

Reply:
Thanks. I'm also not sure how Vampire's allocator feels about deleting a derived class via a pointer of base type so treat that kind of thing with caution. At some point `USE_ALLOCATOR` should go altogether and be retained only for a few select classes.